Add support for `wrapper` option

coreyward opened the pull request:
This is a work-in-progress, and I'd like some feedback. The functionality works, but there aren't any tests yet.

### Wrapper
The `wrapper` option allows for providing a React component or element name that'll be used to create the outermost wrapper around the rendered content, replacing the `div` or `span` that would be used otherwise. 

```jsx
const markdownContent = "# Hello, world!"
const Container = props => <section className="container" {...props} />
const Component = () => (
  <Markdown options={{ wrapper: Container }}>{markdownContent}</Markdown>
)
```
```html
<section class="container">
  <h1>Hello, world!</h1>
</section>
```

This also works with the `compiler` directly.

### Render Array
The `renderArray` option expects a boolean. When `true`, no wrapper will be used at all and an array of elements will be returned. This is especially useful when you want to do some additional processing before handing nodes to React.  This is primarily for use with `compiler`, but can also be useful if you're using HOCs that don't work recursively on Fragment children. 

```jsx
const renderMarkdown = (children, overrides) =>
  compiler(children, {
    renderArray: true,
    forceBlock: true,
    overrides,
  })

// Works like the standard `Markdown` component, but provides an `index`
// prop to each top-level node. Useful for conditional styling.
const Markdown = ({ children }) => (
  <div>
    {renderMarkdown(children, overrides).map((child, index) =>
      React.cloneElement(child, { index })
    )}
  </div>
)
```

coreyward commented:
This would address #113 and #223.

coreyward commented:
@probablyup When do you think you'll have a moment to review this?

probablyup commented:
Hey Corey, perhaps Friday!

On Fri, Sep 27, 2019 at 6:34 PM Corey Ward <notifications@github.com> wrote:

> @probablyup <https://github.com/probablyup> When do you think you'll have
> a moment to review this?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/probablyup/markdown-to-jsx/pull/268?email_source=notifications&email_token=AAELFVRYC25GE3546AN4BADQL2C7VA5CNFSM4IY27LC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD72H5WY#issuecomment-536116955>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAELFVSUWMDWSCOTHND4U23QL2C7VANCNFSM4IY27LCQ>
> .
>

probablyup commented on index.js:
```diff
@@ -773,9 +773,19 @@ export function compiler(markdown, options) {
     if (arr.length > 1) {
-      jsx = inline ? <span key="outer">{arr}</span> : <div key="outer">{arr}</div>;
+      if (options.wrapper) {
+        jsx = React.createElement(options.wrapper, {}, arr)
```
This can be collapsed down to:

```jsx
jsx = React.createElement(options.wrapper || inline ? 'span' : 'div', { key: 'outer' }, arr)
```

probablyup commented:
I'm cool with this, but rather than `renderArray` I'm thinking that passing `null` as the wrapper leads to the wrapperless scenario.

coreyward commented:
Hey @probablyup, I saw #327 looking for this as well and thought I'd pick it back up. I rebased and incorporated the feedback from last year. The new functionality needs tests drafted still, but this is the logic:

- When `wrapper` is provided a non-null value and there are multiple children, the wrapper element is used
- When no `wrapper` is provided, the previous logic of using `div` for block-level or `span` for inline-level elements is used
- When `wrapper` is `null`, an array is returned regardless of the number of children
- Finally, when `wrapper` is provided a non-null value, but there are not multiple children, the prior logic prevails (deferring to custom `createElement` types for single elements, otherwise using the `span` or `div` as above)

Because this still results in cases where it is difficult to reason about the outermost element or component returned by the compiler, I also added a `forceWrapper` option. When this is set, even single element children are wrapped, e.g.:

```js
// Inline
compiler("Hi. This is fun.", { wrapper: "foo", forceWrapper: true })
// => <foo>Hi. This is fun.</foo>

// Block
compiler("Hi. This is fun.", { wrapper: "foo", forceWrapper: true, forceBlock: true })
// => <foo><p>Hi. This is fun.</p></foo>
```

probablyup reviewed (approved):
This looks great to me! Thanks for introducing `theredoc`, looks very cool :)

probablyup commented:
@coreyward Could you submit a quick followup PR with some README docs?
